src/portable/espressif/esp32s2/dcd_esp32s2.c

```diff
@@ -0,0 +1,751 @@
+  xfer->short_packet = (xfer_size < xfer->max_size);
+}
+
+static void transmit_packet(xfer_ctl_t *xfer, volatile usb_in_endpoint_t *in_ep, uint8_t fifo_num)
+{
+  ESP_EARLY_LOGV(TAG, "USB - transmit_packet");
+  volatile uint32_t *tx_fifo = USB0.fifo[fifo_num];
+
+  uint16_t remaining = (in_ep->dieptsiz & 0x7FFFFU) >> USB_D_XFERSIZE0_S;
+  xfer->queued_len = xfer->total_len - remaining;
+
+  uint16_t to_xfer_size = (remaining > xfer->max_size) ? xfer->max_size : remaining;
+  uint8_t to_xfer_rem = to_xfer_size % 4;
+  uint16_t to_xfer_size_aligned = to_xfer_size - to_xfer_rem;
+
+  // Buffer might not be aligned to 32b, so we need to force alignment
+  // by copying to a temp var.
+  uint8_t *base = (xfer->buffer + xfer->queued_len);
+
+  // This for loop always runs at least once- skip if less than 4 bytes
+  // to send off.
+  if (to_xfer_size >= 4) {
+    for (uint16_t i = 0; i < to_xfer_size_aligned; i += 4) {
+      uint32_t tmp = base[i] | (base[i + 1] << 8) |
+          (base[i + 2] << 16) | (base[i + 3] << 24);
+      (*tx_fifo) = tmp;
+    }
+  }
+
+  // Do not read beyond end of buffer if not divisible by 4.
+  if (to_xfer_rem != 0) {
+    uint32_t tmp = 0;
+    uint8_t *last_32b_bound = base + to_xfer_size_aligned;
+
+    tmp |= last_32b_bound[0];
+    if (to_xfer_rem > 1) {
+      tmp |= (last_32b_bound[1] << 8);
+    }
+    if (to_xfer_rem > 2) {
+      tmp |= (last_32b_bound[2] << 16);
+    }
+
+    (*tx_fifo) = tmp;
+  }
+}
+
+static void read_rx_fifo(void)
+{
+  // Pop control word off FIFO (completed xfers will have 2 control words,
+  // we only pop one ctl word each interrupt).
+  uint32_t const ctl_word = USB0.grxstsp;
+  uint8_t  const pktsts   = (ctl_word & USB_PKTSTS_M) >> USB_PKTSTS_S;
+  uint8_t  const epnum    = (ctl_word & USB_CHNUM_M ) >> USB_CHNUM_S;
+  uint16_t const bcnt     = (ctl_word & USB_BCNT_M  ) >> USB_BCNT_S;
+
+  switch (pktsts) {
+    case 0x01: // Global OUT NAK (Interrupt)
+      ESP_EARLY_LOGV(TAG, "TUSB IRQ - RX type : Global OUT NAK");
+      break;
+
+    case 0x02: { // Out packet recvd
+      ESP_EARLY_LOGV(TAG, "TUSB IRQ - RX type : Out packet");
+      xfer_ctl_t *xfer = XFER_CTL_BASE(epnum, TUSB_DIR_OUT);
+      receive_packet(xfer, bcnt);
+    }
+    break;
+
+    case 0x03: // Out packet done (Interrupt)
+      ESP_EARLY_LOGV(TAG, "TUSB IRQ - RX type : Out packet done");
+      break;
+
+    case 0x04: // Step 2: Setup transaction completed (Interrupt)
+      // After this event, OEPINT interrupt will occur with SETUP bit set
+      ESP_EARLY_LOGV(TAG, "TUSB IRQ - RX : Setup packet done");
+      USB0.out_ep_reg[epnum].doeptsiz |= USB_SUPCNT0_M;
+      break;
+
+    case 0x06: { // Step1: Setup data packet received
+      volatile uint32_t *rx_fifo = USB0.fifo[0];
+
+      // We can receive up to three setup packets in succession, but
+      // only the last one is valid. Therefore we just overwrite it
+      _setup_packet[0] = (*rx_fifo);
```

Comment:
@cr1901 I found that the number of setup packet in doeptsiz of esp32 is inconsistent. It can still correctly extract the setup packet and have the SETUP count = 3 !!! I found that just simply overwrite the setup[0-1] with latest from rx_fifo work best and have the behave the same as specs ( process the last setup) still consume less memory 4x2 vs 4x6 of `_setup_packet`

Reply:
@hathach I say do the optimization for now, and if there are setup packet handling issues, revert this change and see if the bug is fixed.

I plan on leaving dcd_synopsys as-is for now

Reply:
@cr1901 on paper it should, but I found out it is not, it still 3 (max) and still received an correct SETUP packet, which cause the racing issue as discussed. OK we will leave the stm32 dcd as it is for now. Should it have some kind of issues with control transfer, we could apply this tip if needed :)

Reply:
@hathach Change the following [line](https://github.com/hathach/tinyusb/pull/319/files#diff-371e8160b188a80c1a3821c75d194adfR334) from:

```
USB0.out_ep_reg[epnum].doeptsiz = USB_PKTCNT0_M |
        ((xfer->max_size & USB_XFERSIZE0_V) << USB_XFERSIZE0_S);
```

to

```
USB0.out_ep_reg[epnum].doeptsiz |= USB_PKTCNT0_M |
        ((xfer->max_size & USB_XFERSIZE0_V) << USB_XFERSIZE0_S);
```

Note the ORing. See if the bug goes away and you can save all 3 setup packets again. This seems quite similar to a [bug](https://github.com/hathach/tinyusb/commit/d4302dacc584c4488cf1905caa822f96bd64930d) I fixed from my original STM32 port.

Reply:
@cr1901 ooppps, I think you mentioned it at the first review. I did a bi-sect but man, that ORing is difficult to spot !!! :eyes: Shame on me. I have a good excuse those, the indent mess up the meld side-by-side comparison. Thanks, I will make an pr fix for this.

Reply:
>I did a bi-sect but man, that ORing is difficult to spot !!! 👀 Shame on me.

No shame at all... I mean, I was the one who wrote the bug in the first place :P. I wrote my initial comments while still sleepy. It took me an hour to realize: "wait... I've seen this before". Anyways I left feedback/requested changes on your PR. I would like to merge the two as well, but hope there's some consistency in register names (or an easy way to define the reg names to vendor-neutral names).

Reply:
Yeah, thanks for pointing it out, we will merge them together later on :).